lua/rip-substitute/matches.lua

```diff
@@ -0,0 +1,221 @@
+	if not closestMatch then
+		for _, match in ipairs(matches) do
+			local on_line_before = match.row < cursor_row
+			local on_same_line_before = match.row == cursor_row and match.col < cursor_col
+
+			if on_line_before or on_same_line_before then
+				closestMatch = match
+				-- No break here; keep updating closestMatch to the last match before cursor
+			end
+		end
+	end
+
+	return closestMatch
+end
+
+---@param replacing boolean
+---@param newSelectedMatchIndex number
+local function updateSelectedMatchHighlight(replacing, newSelectedMatchIndex)
+	local state = require("rip-substitute.state").state
+	vim.api.nvim_buf_clear_namespace(
+		state.targetBuf,
+		state.incPreviewNs,
+		state.selectedMatch.row,
+		state.selectedMatch.row + 1
+	)
+	if replacing then
+		rg.highlightReplacement(state.selectedMatch, false, state.targetBuf, state.incPreviewNs)
+	else
+		rg.highlightMatch(
+			state.selectedMatch,
+			false,
+			state.targetBuf,
+			state.incPreviewNs,
+			state.selectedMatch.col + #state.selectedMatch.matchedText
+		)
+	end
+	state.selectedMatch = state.matches[newSelectedMatchIndex]
+	--TODO: this will probably clear too many highlights
+	vim.api.nvim_buf_clear_namespace(
+		state.targetBuf,
+		state.incPreviewNs,
+		state.selectedMatch.row,
+		state.selectedMatch.row + 1
+	)
+	if replacing then
+		rg.highlightReplacement(state.selectedMatch, false, state.targetBuf, state.incPreviewNs)
+	else
+		rg.highlightMatch(
+			state.selectedMatch,
+			true,
+			state.targetBuf,
+			state.incPreviewNs,
+			state.selectedMatch.col + #state.selectedMatch.matchedText
+		)
+	end
+end
+
+function M.selectPrevMatch()
+	local state = require("rip-substitute.state").state
+	local selectedMatch = state.selectedMatch
+	if not selectedMatch then return end
+	local currentMatchIndex = utils.index_of(
+		state.matches,
+		function(m) return m == selectedMatch end
+	)
+	local prevIndex = -1
+	if currentMatchIndex == 1 then
+		prevIndex = #state.matches
+	else
+		prevIndex = currentMatchIndex - 1
+	end
+	local replacing = selectedMatch.replacementText ~= ""
+	updateSelectedMatchHighlight(replacing, prevIndex)
+	M.centerViewportOnMatch(state.selectedMatch)
+end
+
+function M.selectNextMatch()
+	local state = require("rip-substitute.state").state
+	local selectedMatch = state.selectedMatch
+	if not selectedMatch then return end
+	local replacing = selectedMatch.replacementText ~= ""
+	local currentMatchIndex = utils.index_of(
+		state.matches,
+		function(m) return m == selectedMatch end
+	)
+	local nextIndex = -1
+	if currentMatchIndex == #state.matches then
+		nextIndex = 1
+	else
+		nextIndex = currentMatchIndex + 1
+	end
+	updateSelectedMatchHighlight(replacing, nextIndex)
+	M.centerViewportOnMatch(state.selectedMatch)
+end
+
+---@param match RipSubstituteMatch
+function M.centerViewportOnMatch(match)
+	local state = require("rip-substitute.state").state
+	local padding = 5
```

Comment:
any reason why the padding is 5? Would it maybe make sense to use the user's value for `vim.o.scrolloff` instead?

Reply:
absolutely, great suggestion